### PR TITLE
Laboratory work #1, Maxim Glukhankov - 22FPL1

### PR DESCRIPTION
**Pull request by glukhankov**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
interesting implementation

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -1,68 +1,104 @@
...
 
 def tokenize(text: str) -> list[str] | None:
-    """
-    Splits a text into tokens, converts the tokens into lowercase,
-    removes punctuation, digits and other symbols
-    :param text: a text
-    :return: a list of lower-cased tokens without punctuation
-    """
-
+    if type(text) is str:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:8]: revert the doclines

Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:8]: want to get type - use isinstace

Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:8]: Try early return - already sad on a seminar

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -1,68 +1,104 @@
...
+        return None
 
 def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
-    """
-    Calculates frequencies of given tokens
-    :param tokens: a list of tokens
-    :return: a dictionary with frequencies
-    """
-
+    if type(tokens) is list:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:18]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -1,68 +1,104 @@
...
+        dict_instances = {}
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:19]: tooooooo much code for this small task - please, simplify

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -1,68 +1,104 @@
...
 
 def create_language_profile(language: str, text: str) -> dict[str, str | dict[str, float]] | None:
-    """
-    Creates a language profile
-    :param language: a language
-    :param text: a text
-    :return: a dictionary with two keys – name, freq
-    """
-
+    if type(language) is str and type(text) is str:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:36]: the same - doclines and isinstance

File: wisdom [hunk 5]
```diff
@@ -0,0 +1 @@
+.
```
Comment (artyomtugaryov) [inline wisdom:1]: revert

File: lab_1_classify_by_unigrams/start.py [hunk 6]
```diff
@@ -1,7 +1,7 @@
 """
-Language detection starter
+Language detection starter turi ip ip ip
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:2]: revert

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -37,6 +52,20 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
     :param actual: a list of actual values
     :return: the score
     """
+    if not (isinstance(predicted, list) and isinstance(actual, list)):
+        return None
+    list_sum = []
+    list_squared = []
+    n = 0
+    summa = 0
+    for i in predicted:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:61]: too many loops - merge it

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -63,6 +109,25 @@ def detect_language(
...
+        return None
+    first_pair = compare_profiles(
+        unknown_profile,
+        profile_1
+    )
+    second_pair = compare_profiles(
+        unknown_profile,
+        profile_2
+    )
+    if first_pair < second_pair:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:122]: looks like you have problem with mypy?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py]: done

File: wisdom (hunk 5 above)
Comment (glukhankov) [reply wisdom]: done

File: lab_1_classify_by_unigrams/start.py (hunk 6 above)
Comment (glukhankov) [reply lab_1_classify_by_unigrams/start.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -11,6 +11,13 @@ def tokenize(text: str) -> list[str] | None:
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
     """
+    if not isinstance(text, str):
+        return None
+    tokenized = []
```
[the remark this answers is not in the document]
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -19,6 +26,10 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
+    if not (isinstance(tokens, list) and all(isinstance(token, str) for token in tokens)):
+        return None
+    freq_dict = {token: (tokens.count(token) / len(tokens)) for token in tokens}
```
[the remark this answers is not in the document]
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py]: fixed

**Review by glukhankov — COMMENTED** [review]
all the requests have been adressed

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -2,7 +2,6 @@
 Lab 1
 Language detection
 """
-import json
 
 
 def tokenize(text: str) -> list[str] | None:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:7]: Is it your implementation of this function?

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -61,18 +55,15 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
             and len(predicted) == len(actual)
     ):
         return None
 
-    sum_diff = 0
-    for freq_value in zip(predicted, actual):
-        sum_diff += (freq_value[0] - freq_value[1]) ** 2
-    mse = sum_diff / len(predicted)
-    return mse
+    summa = float(sum((p - a) ** 2 for p, a in zip(predicted, actual)))
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:64]: Why do you cast it to float?

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -127,20 +115,28 @@ def detect_language(
...
+        unknown_profile,
+        profile_1
+    )
+    second_pair = compare_profiles(
+        unknown_profile,
+        profile_2
+    )
+    name_1 = str(profile_1['name'])
+    name_2 = str(profile_2['name'])
+    if isinstance(first_pair, float) and isinstance(second_pair, float):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:128]: revert the condition and fo return as soon as posible

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py (hunk 11 above)
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py:7]: I used to have another implementation (I guess it means the way in which code is realized) of tokenize() which I then simplified using list comprehension.

File: lab_1_classify_by_unigrams/main.py (hunk 12 above)
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py:64]: If I just use sum() without casting its result (summa) to float, then mypy says the return is Optional, and not Float which is required there.

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py:128]: I can't revert it, mypy needs to be sure these are the right types, otherwise it gives errors.

**Review by artyomtugaryov — APPROVED** [review]

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -40,18 +38,14 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
...
-    tokens = tokenize(text)
-    frequency_dict = calculate_frequencies(tokens)
-
-    if not frequency_dict:
+    frequencies = calculate_frequencies(tokenize(text))
+    if not isinstance(frequencies, dict):
         return None
 
-    return {'name': language, 'freq': frequency_dict}
+    return {"name": language, "freq": frequencies}
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:48]: outstanding move

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:128]: I mean you should process cornercases as soon as posible. Return None as soon as posible

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -127,20 +115,28 @@ def detect_language(
...
+        profile_2
+    )
+    name_1 = str(profile_1['name'])
+    name_2 = str(profile_2['name'])
+    if isinstance(first_pair, float) and isinstance(second_pair, float):
+        if first_pair < second_pair:
+            return name_1
+        if second_pair < first_pair:
+            return name_2
+        if first_pair == second_pair:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:133]: so, this you will not need this if

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Looks like nothing new

**Comment by glukhankov** [comment]
Understood.

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (glukhankov) [reply lab_1_classify_by_unigrams/main.py]: done

**Pull request closed without merging**